parsers/pascal.c

```diff
@@ -74,6 +80,68 @@ static bool tail (const char *cp)
 	return result;
 }
 
+static void parseArglist(const char *buf, char **arglist, char **vartype)
+{
+	char *c, *start, *end;
+	int level;
+
+	if (NULL == buf || NULL == arglist)
+		return;
+
+	c = strdup(buf);
+	/* parse argument list which can be missing like in "function ginit:integer;" */
+	if (NULL != (start = strchr(c, '(')))
+	{
+		for (level = 1, end = start + 1; level > 0; ++end)
+		{
+			if ('\0' == *end)
+				break;
+			else if ('(' == *end)
+				++ level;
+			else if (')' == *end)
+				-- level;
+		}
+	}
+	else /* if no argument list was found, continue looking for a return value */
+	{
+		start = "()";
+		end = c;
+	}
+
+	/* parse return type if requested by passing a non-NULL vartype argument */
+	if (NULL != vartype)
+	{
+		char *var, *var_start;
+
+		*vartype = NULL;
+
+		if (NULL != (var = strchr(end, ':')))
+		{
+			var++; /* skip ':' */
+			while (isspace((int) *var))
+				++var;
+
+			if (starttoken(*var))
+			{
+				var_start = var;
+				var++;
+				while (intoken(*var))
+					var++;
+				if (endtoken(*var))
+				{
+					*var = '\0';
+					*vartype = strdup(var_start);
+				}
+			}
+		}
+	}
+
+	*end = '\0';
+	*arglist = strdup(start);
```

Comment:
My understanding:

To do `*end = '\0'`, you have to do `c = strdup(buf)`.
To do `strdup(start)`, you have to do `*end = '\0'`, terminating the C string.

I think you can reduce `strdup()` with `eStrndup(const char* str, size_t len);` declared in `main/routines.h`.
With it, you can make a new C string without doing `*end = '\0'`.

Reply:
Thanks for the hint, did it and then rewrote it with your `vString` suggestion below :D.